ext/java/org/msgpack/jruby/Packer.java

```diff
@@ -30,20 +35,139 @@ public IRubyObject allocate(Ruby runtime, RubyClass type) {
+    private Ruby runtime;
+    public RubyHash hash;
+    public RubyHash cache;
+
+    public ExtRegistry(Ruby runtime) {
+      this.runtime = runtime;
+      hash = RubyHash.newHash(runtime);
+      cache = RubyHash.newHash(runtime);
+    }
+
+    public ExtRegistry dup() {
+      ExtRegistry copy = new ExtRegistry(this.runtime);
+      copy.hash = (RubyHash) this.hash.dup(runtime.getCurrentContext());
+      copy.cache = RubyHash.newHash(runtime);
+      return copy;
+    }
+
+    public void put(RubyClass klass, int typeId, IRubyObject proc, IRubyObject arg) {
+      RubyArray e = RubyArray.newArray(runtime, new IRubyObject[] { RubyFixnum.newFixnum(runtime, typeId), proc, arg });
+      cache.rb_clear();
+      hash.fastASet(klass, e);
+    }
+
+    // proc, typeId(Fixnum)
+    public IRubyObject[] lookup(final RubyClass klass) {
+      RubyArray e = (RubyArray) hash.fastARef(klass);
+      if (e == null) {
+        e = (RubyArray) cache.fastARef(klass);
+      }
+      if (e != null) {
+        IRubyObject[] hit = new IRubyObject[] {};
+        hit[0] = e.entry(1);
+        hit[1] = e.entry(0);
+        return hit;
+      }
+
+      final IRubyObject[] pair = new IRubyObject[2];
+      // check all keys whether it's super class of klass, or not
+      hash.visitAll(new RubyHash.Visitor() {
+        public void visit(IRubyObject keyValue, IRubyObject value) {
+          if (pair[0] != null) {
+            return;
+          }
+          RubyClass key = (RubyClass) keyValue;
+          IRubyObject rb_class_inherited_p = klass.include_p(runtime.getCurrentContext(), key);
+          if (rb_class_inherited_p.isTrue()) {
+            RubyArray hit = (RubyArray) hash.fastARef(keyValue);
+            cache.fastASet(klass, hit);
+            pair[0] = hit.entry(1);
+            pair[1] = hit.entry(0);
+          }
+        }
+      });
+
+      if (pair[0] == null) {
+        return null;
+      }
+      return pair;
+    }
+  }
+
   @JRubyMethod(name = "initialize", optional = 2)
   public IRubyObject initialize(ThreadContext ctx, IRubyObject[] args) {
     boolean compatibilityMode = false;
     if (args.length > 0 && args[args.length - 1] instanceof RubyHash) {
       RubyHash options = (RubyHash) args[args.length - 1];
-      compatibilityMode = options.fastARef(ctx.getRuntime().newSymbol("compatibility_mode")).isTrue();
+      IRubyObject mode = options.fastARef(ctx.getRuntime().newSymbol("compatibility_mode"));
+      compatibilityMode = (mode != null) && mode.isTrue();
     }
     this.encoder = new Encoder(ctx.getRuntime(), compatibilityMode);
     this.buffer = new Buffer(ctx.getRuntime(), ctx.getRuntime().getModule("MessagePack").getClass("Buffer"));
     this.buffer.initialize(ctx, args);
+    this.registry = new ExtRegistry(ctx.getRuntime());
     return this;
   }
 
-  @JRubyMethod(name = "write")
+  public void setExtRegistry(ExtRegistry registry) {
+    this.registry = registry;
+    this.encoder.setRegistry(registry);
+  }
```

Comment:
This method should not be needed, the registry should be injected in the constructor.